ibtk/src/math/PETScMatUtilities.cpp

```diff
@@ -806,10 +596,731 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
+                                                              IS** is_nonoverlap,
+                                                              int& n_subdomains,
+                                                              const IntVector<NDIM>& box_size,
+                                                              const IntVector<NDIM>& overlap_size,
+                                                              const std::vector<int>& num_dofs_per_proc,
+                                                              int dof_index_idx,
+                                                              Pointer<PatchLevel<NDIM> > patch_level)
+{
+    // Determine the DOF index range on this processor.
+    const int mpi_rank = SAMRAI_MPI::getRank();
+    const int dofs_local = num_dofs_per_proc[mpi_rank];
+    const int dof_lower = std::accumulate(num_dofs_per_proc.begin(), num_dofs_per_proc.begin() + mpi_rank, 0);
+    const int dof_upper = dof_lower + dofs_local;
+
+    // Determine the total nonoverlapping subdomains on this processor.
+    n_subdomains = 0;
+    for (PatchLevel<NDIM>::Iterator p(patch_level); p; p++)
+    {
+        Pointer<Patch<NDIM> > patch = patch_level->getPatch(p());
+        const Box<NDIM>& patch_box = patch->getBox();
+        const IntVector<NDIM>& patch_lower = patch_box.lower();
+        const IntVector<NDIM>& patch_upper = patch_box.upper();
+
+        // Compute number of nonoverlapping subdomains
+        IntVector<NDIM> cells = 1;
+        cells += patch_upper - patch_lower;
+        IntVector<NDIM> subdomains = cells / box_size;
+        int n_patch_subdomains = 1;
+        for (int d = 0; d < NDIM; ++d)
+        {
+            // If box size is bigger than patch size, we still account for it.
+            if (subdomains(d) == 0)
+            {
+                subdomains(d) = 1;
+            }
+            n_patch_subdomains *= subdomains(d);
+        }
+        n_subdomains += n_patch_subdomains;
+    }
+    *is_overlap = new IS[n_subdomains];
+    *is_nonoverlap = new IS[n_subdomains];
+
+    // Fill in the IS'es
+    int counter = 0;
+    bool there_is_overlap = overlap_size.min();
+    for (PatchLevel<NDIM>::Iterator p(patch_level); p; p++)
+    {
+        Pointer<Patch<NDIM> > patch = patch_level->getPatch(p());
+        const Box<NDIM>& patch_box = patch->getBox();
+        const IntVector<NDIM>& patch_lower = patch_box.lower();
+        const IntVector<NDIM>& patch_upper = patch_box.upper();
+        Pointer<CellData<NDIM, int> > dof_data = patch->getPatchData(dof_index_idx);
+        const int data_depth = dof_data->getDepth();
+
+        // Compute number of nonoverlapping subdomains
+        IntVector<NDIM> cells = 1;
+        cells += patch_upper - patch_lower;
+        IntVector<NDIM> subdomains = cells / box_size;
+        for (int d = 0; d < NDIM; ++d)
+        {
+            if (subdomains(d) == 0)
+            {
+                subdomains(d) = 1;
+            }
+        }
+
+#if (NDIM == 2)
```

Comment:
One thing that you could do is to make a helper function that will actually do the basic partitioning for you --- e.g. pass in a patch Box, a subdomain size, and a overlap size, and it will return the interior and overlap boxes. I am 99% sure that this can be done in a NDIM-independent manner.